src/saveload/cargopacket_sl.cpp

```diff
@@ -76,6 +76,53 @@
 	if (IsSavegameVersionBefore(SLV_181)) {
 		for (Vehicle *v : Vehicle::Iterate()) v->cargo.KeepAll();
 	}
+
+	if (IsSavegameVersionBefore(SLV_CARGO_TRAVELLED)) {
+		for (Station *st : Station::Iterate()) {
+			for (size_t i = 0; i < NUM_CARGO; i++) {
+				GoodsEntry *ge = &st->goods[i];
+				for (auto it = ge->cargo.Packets()->begin(); it != ge->cargo.Packets()->end(); it++) {
+					for (CargoPacket *cp : it->second) {
+						if (cp->source_xy != INVALID_TILE) {
+							/* For traveled distance, estimate it as the distance of the current station from the source. */
+							cp->travelled.x = TileX(cp->source_xy) - TileX(st->xy);
+							cp->travelled.y = TileY(cp->source_xy) - TileY(st->xy);
+						}
+					}
+				}
+			}
+		}
+
+		/* Find a reasonable source and travel distance for cargo in vehicles. */
+		for (Vehicle *v : Vehicle::Iterate()) {
+			for (auto it = v->cargo.Packets()->begin(); it != v->cargo.Packets()->end(); it++) {
+				/* Assume the cargo was loaded at the source. */
+				if ((*it)->source_xy != INVALID_TILE) {
+					(*it)->UpdateLoadingTile((*it)->source_xy);
+					continue;
+				}
+
+				/* Older games don't have a source location; base it on the station location instead. */
+				auto first_station_id = (*it)->first_station;
+				if (first_station_id != INVALID_STATION) {
+					auto first_station = Station::Get(first_station_id);
+					(*it)->UpdateLoadingTile(first_station->xy);
+					continue;
+				}
+
+				/* And some games also don't have a first-station. Act like it loaded at the current tile.
+				 * This is always wrong, but the best we can do. */
+				(*it)->UpdateLoadingTile(v->tile);
```

Comment:
According to some comments in the source `Vehicle::tile` is zero for aircraft in flight. `Vehicle::x_pos` is kind of valid in flight, but may be outside of the map, if aircraft is in landing-pattern of an airport near the map border.

Reply:
Hmm.. for some reason I missed some earlier code, that already restores `source_xy` for old games. And that already uses `Vehicle::tile`. So this has been broken for a long time for aircrafts.

Mind you, this is only a problem for really old (think: TTO-old) games. The profit they make in their first leg will be weird compared to what it should pay. I am not actually sure it is worth writing a lot of code to prevent that, and maybe fail. I cannot actually test any of it :P

The reason I did do another cascade here of checking if the `source_xy` and `first_station` are valid, as I am not actually sure if anything between `SLV_44` and now broke any of this. But maybe we could just say: fuck that, let's assume `source_xy` is already restored in previous code, as before this PR it is also assumed to be valid.